tree/tree/inc/TBasket.h

```diff
@@ -45,19 +44,30 @@ class TBasket : public TKey {
    // Helper for managing the compressed buffer.
    void InitializeCompressedBuffer(Int_t len, TFile* file);
 
+   // Handles special logic around deleting / reseting the entry offset pointer.
+   void ResetEntryOffset();
+
+   // Get entry offset as result of a calculation.
+   Int_t *GetCalculatedEntryOffset();
+
+   // Returns true if the underlying TLeaf can regenerate the entry offsets for us.
+   Bool_t CanGenerateOffsetArray();
+
 protected:
-   Int_t       fBufferSize;      ///< fBuffer length in bytes
-   Int_t       fNevBufSize;      ///< Length in Int_t of fEntryOffset OR fixed length of each entry if fEntryOffset is null!
-   Int_t       fNevBuf;          ///< Number of entries in basket
-   Int_t       fLast;            ///< Pointer to last used byte in basket
-   Bool_t      fHeaderOnly;      ///< True when only the basket header must be read/written
-   UChar_t     fIOBits{0};       ///<!IO feature flags.  Serialized in custom portion of streamer to avoid forward compat issues unless needed.
-   Int_t      *fDisplacement;    ///<![fNevBuf] Displacement of entries in fBuffer(TKey)
-   Int_t      *fEntryOffset;     ///<[fNevBuf] Offset of entries in fBuffer(TKey)
-   TBranch    *fBranch;          ///<Pointer to the basket support branch
-   TBuffer    *fCompressedBufferRef; ///<! Compressed buffer.
-   Bool_t      fOwnsCompressedBuffer; ///<! Whether or not we own the compressed buffer.
-   Int_t       fLastWriteBufferSize; ///<! Size of the buffer last time we wrote it to disk
+   Int_t       fBufferSize{0}; ///< fBuffer length in bytes
+   Int_t       fNevBufSize{0}; ///< Length in Int_t of fEntryOffset OR fixed length of each entry if fEntryOffset is null!
+   Int_t       fNevBuf{0};     ///< Number of entries in basket
+   Int_t       fLast{0};       ///< Pointer to last used byte in basket
+   Bool_t      fHeaderOnly{kFALSE}; ///< True when only the basket header must be read/written
+   UChar_t     fIOBits{0}; ///<!IO feature flags.  Serialized in custom portion of streamer to avoid forward compat issues unless needed.
+   Bool_t      fReadEntryOffset{kFALSE}; ///<!Set to true if offset array was read from a file.
+   Int_t      *fDisplacement{nullptr};   ///<![fNevBuf] Displacement of entries in fBuffer(TKey)
+   Int_t      *fEntryOffset{nullptr}; ///<[fNevBuf] Offset of entries in fBuffer(TKey); generated at runtime.  Special value
+                                 /// of `-1` indicates that the offset generation MUST be performed on first read.
+   TBranch    *fBranch{nullptr};    ///<Pointer to the basket support branch
+   TBuffer    *fCompressedBufferRef{nullptr}; ///<! Compressed buffer.
+   Bool_t      fOwnsCompressedBuffer{kFALSE};   ///<! Whether or not we own the compressed buffer.
```

Comment:
try moving this to just after fReadEntryOffset and setting them both to be fields:
```
Bool_t      fReadEntryOffset{kFALSE} : 1; ///<!Set to true if offset array was read from a file.
Bool_t      fOwnsCompressedBuffer{kFALSE} : 1;   ///<! Whether or not we own the compressed buffer.
```